[#1] 로그인 기능  패키지 구성/ 회원 가입 구현

jieun-dev1 opened the pull request:
[0829]
아직 기능 완성  전이고 로그인 기능에 문제가 있으나, 미팅 때 논의해보면 좋을 것 같아서 아직 미완성 임에도 푸시합니다.

[0902]
[변경 사항]
-SessionManager로 세션 구현 
-인터셉터 적용 

*테스트 내역 
-> 회원 가입, 로그아웃, 로그인 이후 세션이 유지됨을 확인했습니다 (ex. 홈 화면에서 handler 를 true 로 반환) 

[리뷰/피드백 필요한 내용] 
(1) 로그인 되지 않은 상태로 요청할 경우(테스트: users/home ) error 로 redirect 되면서, 10번씩 로그가 찍힙니다. 
에러인 경우에 이렇게 반복 되는데, 디버깅을 하지 못해서 우선은 올립니다. 로그인 한 경우에는 preHandle이 true 로 잘 리턴 되는데요. 어떤 부분이 문제일까요? 
(2) user logout/expireSession  기능은 한 번 처리를 완료하고도, 계속 누를 수 있더라구요. 한번 삭제 하고 나면, sessionStore에서 삭제 되어서 찾을 수 없을 줄 알았는데, 삭제가 제대로 되지 않은 걸까요?

[추후 작업 필요] 
-Argument Resolver 적용 

[0907]
[변경 사항]
-ExceptionHandler와 ControllerAdvice를 적용하여 JSON 형태로 예외가 반환되도록 했습니다. 
-가입 시 Role 을 입력하도록 추가했습니다. 
-SignUpCommand 를 immutable 하게 유지하고, 레이어드 아키텍처를 유지하기 위해서 서비스 로직을 변경하였습니다. 
-@LoginCheck 로 로그인 인증이 필요한 메서드 블럭만 적용하였습니다. 
(기존의 PathPattern은 url 오타 등 실수로 인한 에러의 여지가 있기 때문입니다)
-checkstyle을 적용하여 일관된 코드 스타일을 구현할 수 있도록 했습니다. (Missing Javadoc 등 불필요한 체크는 추후 suppressions 을 적용할 예정입니다. 현재는 아직 안되고 있습니다)

jieun-dev1 committed:
User 클래스 생성

jieun-dev1 committed:
Feat: 회원 가입 기능 구현

jieun-dev1 committed:
Feat: 로그인 기능 구현

f-lab-bright commented on src/main/java/com/example/freshCart/presentation/UserController.java:
```diff
@@ -0,0 +1,58 @@
+    회원 가입
+     */
+    @PostMapping("/signup")
+    public void Signup(@RequestBody SignupRequest request){
```
일반적으로 메소드 이름은 소문자로 시작하는게 관례라고 생각하시면 됩니다.

f-lab-bright commented on src/main/java/com/example/freshCart/application/UserRegisterServiceV1.java:
```diff
@@ -0,0 +1,47 @@
+        중복체크를 별도 API 로 처리하더라도, 중복일 경우 가입 자체가 되지 않도록 예외 처리
+         */
+    @Override
+    public void toRegister(SignupRequest request) {
```
메소드 명이 왜 toRegister 인가요..?
register 라는 이름으로 충분하지 않나요?
toRegister 라고 하면 뭔가 request를 이용해서 변환된 값이 반환될 것 같은 이름이네요.

jieun-dev1 replied:
그렇네요. 수정했습니다 :)

f-lab-bright commented on src/main/java/com/example/freshCart/application/UserRegisterServiceV1.java:
```diff
@@ -0,0 +1,47 @@
+import com.example.freshCart.domain.User;
+import com.example.freshCart.domain.UserRegisterService;
+import com.example.freshCart.domain.UserRepository;
+import com.example.freshCart.presentation.request.EmailDupCheckRequest;
+import com.example.freshCart.presentation.request.LoginRequest;
+import com.example.freshCart.presentation.request.SignupRequest;
```
presentation layer의 dto들이 application layer에서 쓰이고 있네요.
이건 의존성의 방향이 잘못된 케이스라고 볼 수 있겠네요.
layered architecture에서의 의존성의 방향이 어떻게 향해야 하는지 한 번 고민해보고 수정해보면 좋을 것 같습니다 :)

jieun-dev1 replied:
이번 PR 에서 수정해보았습니다 Application 패키지로 옮겼고, repository들이 이 객체를 의존하지 않도록 했습니다. 검토부탁드려요!

jieun-dev1 committed:
Merge branch 'main' into feature/2-login

jieun-dev1 committed:
Feat: 레이어드 아키텍처에 맞게 Request 객체들을 비즈니스 로직에서 활용하는 command 객체로 재수정.

jieun-dev1 committed:
Feat: Session을 직접 구현하기 위해 SessionUtil 은 우선 주석처리해둠.
LoginUser - 세션 유지 시 사용할 로그인 된 유저객체

jieun-dev1 committed:
Feat: 예외처리

jieun-dev1 committed:
Feat: 세션 추가 확인에 필요한 API 들 추가.

jieun-dev1 committed:
Feat: Request객체를 command 로 수정함

jieun-dev1 committed:
Feat: 인터셉터 구현 (ArgumentResolver 아직 미적용)

jieun-dev1 committed:
Feat: Domain 이 Application 에 의존하는 경우 수정

jieun-dev1 committed:
Feat: 주석, 패키지 구조 변경 등 소소한 수정
UserController에 logout과 home 등 적용
AuthenticationService 삭제

jieun-dev1 committed:
Feat: 포맷 변경에 따른 변경사항 반영

f-lab-bright commented on build.gradle:
```diff
@@ -28,6 +28,8 @@ dependencies {
 	annotationProcessor 'org.projectlombok:lombok'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+//	implementation 'org.mindrot:jbcrypt:0.4'
+	implementation group: 'org.mindrot', name: 'jbcrypt', version: '0.4'
```
jbcrypt를 직접 써도 되지만, `org.springframework.security:spring-security-crypto` 요 의존성을 사용해서 구현해도 좋습니다 :)

jieun-dev1 replied:
구현 할 때, 스프링 시큐리티 프레임워크를 쓰지 않으려고 주의했습니다. 
그래서 bcrypt 만 가능한 줄 알았네요! 해당 의존성은 spring security 프레임워크로 보여서 쓰면 안될 줄 알았는데,  spring-boot-starter-security 와는 또 다른 라이브러리 인가 봅니다!

f-lab-bright replied:
네 요녀석은 spring-security가 사용하는 암호화 관련된 것들만 가지고 있는 의존성이라 생각하시면 됩니다 :)

f-lab-bright commented on src/main/java/com/example/freshCart/application/command/SignupCommand.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+  // 암호화된 패스워드로 바꾸기 BCrytpEncoder에서 가져와야 함.
+  public void toEncryptPassword(String encryptedPassword) {
+        this.password = encryptedPassword;
+    }
```
이걸 dto의 값을 바꾸는게 적절한 방법일까요..?
getter만 제공되는 dto의 경우에는 가능하면 `immutable`하게 사용하는게 좋습니다.

f-lab-bright commented on src/main/java/com/example/freshCart/presentation/web/LoginCheckInterceptor.java:
```diff
@@ -0,0 +1,42 @@
+    LoginUser user = sessionManager.getSession(request);
+    if (user == null) {
+      log.info("인증되지 않은 요청" + requestURI);
+      response.sendRedirect("/users/login?redirectURL=" + requestURI);
```
api 동작이기 때문에 redirect를 할 필요 없고, 인증/인가에 관련된 예외 응답을 내려주시면 됩니다 :)

jieun-dev1 replied:
만약 view 가 있는 프로젝트라면, 이렇게 redirect 가 맞겠지만 생각해보니 api 는 그럴 필요가 전혀 없네요 ㅎㅎ 수정하였습니다.

f-lab-bright commented on src/main/java/com/example/freshCart/presentation/web/WebConfig.java:
```diff
@@ -0,0 +1,32 @@
+        .addInterceptor(new LoginCheckInterceptor(sessionManager))
+        .order(1)
+        .addPathPatterns("/**")
+        .excludePathPatterns("/users/login", "/users/logout", "/users/signup");
```
인가 처리가 필요하지 않은 api가 생길 때 마다 여기 패턴을 계속 추가해줘야 하겠네요 :)
여기 코드를 계속 추가하지 않고 할 수 있는 방법은 없을까요??

jieun-dev1 replied:
혹시 키워드를 알려주실 수 있을까요? 이 부분 계속 고민하고 찾아보고 있는데, 아직 실마리를 못 찾았습니다.

f-lab-bright replied:
인터셉터의 메소드를 보면 컨트롤러의 메소드의 정보를 읽어올 수 있는데, 이 때 어노테이션의 정보도 읽어올 수 있습니다 :)
이를 활용해볼 수도 있지 않을까요? :)

jieun-dev1 replied:
네 @LoginCheck 로 로그인 인증이 필요한 메서드 블럭만 적용하였습니다. @AuthenticatedUser 로 로그인한 유저의 정보를 파라미터에서 적용할 수 있도록 했습니다.

jieun-dev1 committed:
Feat: 회원 가입 시 필수 값 입력을 강제하도록 @NotNull과 Valid 사용

jieun-dev1 committed:
Feat: 예외 처리

사용자 정의 Exception 클래스들이 BaseException 클래스를 상속받아, 핸들러에서 공통으로 예외 처리

jieun-dev1 committed:
Chore: 포맷터 적용에 따른 변경

jieun-dev1 committed:
Feat: NonNull을 User객체가 아닌 SignUpCommand에 적용

jieun-dev1 committed:
Feat: 회원 가입 로직 구현 변경

command 객체는 Immutable하게 유지할 수 있도록, User객체를 만들 때 패스워드 암호화를 적용함

 emailDupCheck는 서비스 로직 내에서 검사하는 것으로 충분하여 삭제함

jieun-dev1 committed:
Feat: Unexpected 에러를 검사하기 위해 users/unexpected를 exclude Path처리함.

Exception 에 String을 인자로 넣지 않고 클래스 내에서 static message로 생성

jieun-dev1 committed:
Feat: Path Pattern 대신 어노테이션을 적용

-Path 변경 시 생길 수 있는 에러를 줄일 수 있음
-로그인이 필요한 로직에 대해 컨트롤러 클래스에서 바로 확인 가능

jieun-dev1 committed:
Feat: Path Pattern 대신 어노테이션을 적용

-Path 변경 시 생길 수 있는 에러를 줄일 수 있음
-로그인이 필요한 로직에 대해 컨트롤러 클래스에서 바로 확인 가능

jieun-dev1 committed:
Chore: 주석 추가와 패키지명 소문자 변경

jieun-dev1 committed:
Chore: 주석 추가와 패키지명 소문자 변경

jieun-dev1 committed:
Chore: 패키지 명 소문자로 변경
자바 상수 규약에 맞게 대문자와 카멜케이스로 변경
